File: backend-side/app/channels/chat_room_channel.rb
class ChatRoomChannel < ApplicationCable::Channel
  def subscribed
    stream_from "chat_room_#{params[:room_id]}"
  end

  def unsubscribed
    stop_all_streams
  end

  def speak(data)
    message = Message.create!(
      chat_room_id: params[:room_id],
      content: data['content']
    )
    
    # The broadcast is now handled by the Message model's after_create_commit callback
  end
end

File: backend-side/app/javascript/channels/chat_room_channel.js
import consumer from "channels/consumer"

export default function subscribeToRoom(roomId, receivedCallback) {
  return consumer.subscriptions.create(
    { channel: "ChatRoomChannel", room: roomId },
    {
      received(data) {
        receivedCallback(data)
      }
    }
  )
}
